File: nativeshell_dart/lib/src/api_constants.dart
class Channels {
  static final dispatcher = 'nativeshell/window-dispatcher';

  // window sub channels
  static final windowManager = '.window.window-manager';
  static final dropTarget = '.window.drop-target';
  static final dragSource = '.window.drag-source';

  static final menuManager = 'nativeshell/menu-manager';
  static final keyboardMapManager = 'nativeshell/keyboard-map-manager';
  static final hotKeyManager = 'nativeshell/hot-key-manager';
  static final screenManager = 'nativeshell/screen-manager';
  static final statusItemManager = 'nativeshell/status-item-manager';
}

class Events {
  static final windowInitialize = 'event:Window.initialize';
  static final windowVisibilityChanged = 'event:Window.visibilityChanged';
  static final WindowStateFlagsChanged = 'event:Window.stateFlagsChanged';
  static final WindowGeometryChanged = 'event:Window.geometryChanged';
  static final windowCloseRequest = 'event:Window.closeRequest';
  static final windowClose = 'event:Window.close';
}

const currentApiVersion = 1;

class Methods {
  // WindowManager
  static final windowManagerGetApiVersion = 'WindowManager.getApiVersion';
  static final windowManagerCreateWindow = 'WindowManager.createWindow';
  static final windowManagerInitWindow = 'WindowManager.initWindow';

  // Window
  static final windowShow = 'Window.show';
  static final windowShowModal = 'Window.showModal';
  static final windowReadyToShow = 'Window.readyToShow';
  static final windowHide = 'Window.hide';
  static final windowActivate = 'Window.activate';
  static final windowDeactivate = 'Window.deactivate';
  static final windowClose = 'Window.close';
  static final windowCloseWithResult = 'Window.closeWithResult';

  static final windowSetGeometry = 'Window.setGeometry';
  static final windowGetGeometry = 'Window.getGeometry';
  static final windowSupportedGeometry = 'Window.supportedGeometry';
  static final windowGetScreenId = 'Window.getScreenId';

  static final windowSetStyle = 'Window.setStyle';
  static final windowSetTitle = 'Window.setTitle';
  static final windowSetMinimized = 'Window.setMinimized';
  static final windowSetMaximized = 'Window.setMaximized';
  static final windowSetFullScreen = 'Window.setFullScreen';
  static final windowSetCollectionBehavior = 'Window.setCollectionBehavior';
  static final windowGetWindowStateFlags = 'Window.getWindowStateFlags';
  static final windowPerformWindowDrag = 'Window.performWindowDrag';

  static final windowShowPopupMenu = 'Window.showPopupMenu';
  static final windowHidePopupMenu = 'Window.hidePopupMenu';
  static final windowShowSystemMenu = 'Window.showSystemMenu';
  static final windowSetWindowMenu = 'Window.setWindowMenu';

  static final windowSavePositionToString = 'Window.savePositionToString';
  static final windowRestorePositionFromString =
      'Window.restorePositionFromString';

  // Drag Driver
  static final dragDriverDraggingUpdated = 'DragDriver.draggingUpdated';
  static final dragDriverDraggingExited = 'DragDriver.draggingExited';
  static final dragDriverPerformDrop = 'DragDriver.performDrop';

  // Drop Source
  static final dragSourceBeginDragSession = 'DragSource.beginDragSession';
  static final dragSourceDragSessionEnded = 'DragSource.dragSessionEnded';

  // Menu
  static final menuCreateOrUpdate = 'Menu.createOrUpdate';
  static final menuDestroy = 'Menu.destroy';
  static final menuOnAction = 'Menu.onAction';
  static final menuOnOpen = 'Menu.onOpen';
  static final menuSetAppMenu = 'Menu.setAppMenu';

  // Menubar
  static final menubarMoveToPreviousMenu = 'Menubar.moveToPreviousMenu';
  static final menubarMoveToNextMenu = 'Menubar.moveToNextMenu';

  // KeyboardMap
  static final keyboardMapGet = 'KeyboardMap.get';
  static final keyboardMapOnChanged = 'KeyboardMap.onChanged';

  // HotKey
  static final hotKeyCreate = 'HotKey.create';
  static final hotKeyDestroy = 'HotKey.destroy';
  static final hotKeyOnPressed = 'HotKey.onPressed';

  // ScreenManager
  static final screenManagerScreensChanged = 'ScreenManager.screensChanged';
  static final screenManagerGetScreens = 'ScreenManager.getScreens';
  static final screenManagerGetMainScreen = 'ScreenManager.getMainScreen';
  static final screenManagerLogicalToSystem = 'ScreenManager.logicalToSystem';
  static final screenManagerSystemToLogical = 'ScreenManager.systemToLogical';

  // StatusItem
  static final statusItemInit = 'StatusItem.init';
  static final statusItemCreate = 'StatusItem.create';
  static final statusItemDestroy = 'StatusItem.destroy';
  static final statusItemSetImage = 'StatusItem.setImage';
  static final statusItemSetHint = 'StatusItem.setHint';
  static final statusItemShowMenu = 'StatusItem.showMenu';
  static final statusItemSetHighlighted = 'StatusItem.setHighlighted';
  static final statusItemGetGeometry = 'StatusItem.getGeometry';
  static final statusItemGetScreenId = 'StatusItem.getScreenId';
  static final statusItemOnAction = 'StatusItem.onAction';
}

class Keys {
  static final dragDataFiles = 'drag-data:internal:files';
  static final dragDataURLs = 'drag-data:internal:urls';
}

File: nativeshell/src/shell/api_constants.rs
pub(crate) mod channel {
    // Flutter channel for windows bound messages; All messages that concern windows are
    // dispatched on this channel
    pub const DISPATCHER: &str = "nativeshell/window-dispatcher";

    // Window sub channels (delivered by dispatcher)
    pub mod win {
        pub const WINDOW_MANAGER: &str = ".window.window-manager";
        pub const DROP_TARGET: &str = ".window.drop-target";
        pub const DRAG_SOURCE: &str = ".window.drag-source";
    }

    // Flutter channel for managing platform menus
    pub const MENU_MANAGER: &str = "nativeshell/menu-manager";

    // Flutter channel for keyboard layout notifications
    pub const KEYBOARD_MAP_MANAGER: &str = "nativeshell/keyboard-map-manager";

    // Flutter channel for managing hot keys
    pub const HOT_KEY_MANAGER: &str = "nativeshell/hot-key-manager";
}

pub const CURRENT_API_VERSION: i32 = 1;

pub(crate) mod method {

    pub mod window_manager {
        pub const GET_API_VERSION: &str = "WindowManager.getApiVersion";

        // Request creation of new window
        pub const CREATE_WINDOW: &str = "WindowManager.createWindow";

        // Initializes current isolate window
        pub const INIT_WINDOW: &str = "WindowManager.initWindow";
    }

    pub mod window {

        // Request to show the window (may be delayed until window itself calls readyToShow)
        pub const SHOW: &str = "Window.show";

        // Request to show the window modally (will return result after window closes)
        pub const SHOW_MODAL: &str = "Window.showModal";

        // Called by window itself after the layout is ready and window is prepared to be shown
        pub const READY_TO_SHOW: &str = "Window.readyToShow";

        // Hide the window
        pub const HIDE: &str = "Window.hide";

        // Bring window front and request focus
        pub const ACTIVATE: &str = "Window.activate";

        // Close the window; This will terminate the isolate
        pub const CLOSE: &str = "Window.close";

        pub const CLOSE_WITH_RESULT: &str = "Window.closeWithResult";

        // All positions, unless otherwise noted are in logical coordinates with top left origin

        pub const SET_GEOMETRY: &str = "Window.setGeometry";
        pub const GET_GEOMETRY: &str = "Window.getGeometry";
        pub const SUPPORTED_GEOMETRY: &str = "Window.supportedGeometry";

        pub const SET_STYLE: &str = "Window.setStyle";
        pub const SET_TITLE: &str = "Window.setTitle";
        pub const PERFORM_WINDOW_DRAG: &str = "Window.performWindowDrag";

        pub const SHOW_POPUP_MENU: &str = "Window.showPopupMenu";
        pub const HIDE_POPUP_MENU: &str = "Window.hidePopupMenu";

        // Windows only
        pub const SHOW_SYSTEM_MENU: &str = "Window.showSystemMenu";

        // MacOS only - associates given menu with current windon; The menu will
        // be displayed  when window gets active
        pub const SET_WINDOW_MENU: &str = "Window.setWindowMenu";

        pub const SAVE_POSITION_TO_STRING: &str = "Window.savePositionToString";
        pub const RESTORE_POSITION_FROM_STRING: &str = "Window.restorePositionFromString";
    }

    pub mod drag_driver {
        pub const DRAGGING_UPDATED: &str = "DragDriver.draggingUpdated";
        pub const DRAGGING_EXITED: &str = "DragDriver.draggingExited";
        pub const PERFORM_DROP: &str = "DragDriver.performDrop";
    }

    pub mod drag_source {
        pub const BEGIN_DRAG_SESSION: &str = "DragSource.beginDragSession";
        pub const DRAG_SESSION_ENDED: &str = "DragSource.dragSessionEnded";
    }

    pub mod menu {
        pub const CREATE_OR_UPDATE: &str = "Menu.createOrUpdate";
        pub const DESTROY: &str = "Menu.destroy";
        pub const ON_ACTION: &str = "Menu.onAction";
        pub const ON_OPEN: &str = "Menu.onOpen";
        pub const SET_APP_MENU: &str = "Menu.setAppMenu";
    }

    pub mod menu_bar {
        // Menubar - move to previous menu
        pub const MOVE_TO_PREVIOUS_MENU: &str = "Menubar.moveToPreviousMenu";
        pub const MOVE_TO_NEXT_MENU: &str = "Menubar.moveToNextMenu";
    }

    pub mod keyboard_map {
        pub const GET: &str = "KeyboardMap.get";
        pub const ON_CHANGED: &str = "KeyboardMap.onChanged";
    }

    pub mod hot_key {
        pub const CREATE: &str = "HotKey.create";
        pub const DESTROY: &str = "HotKey.destroy";
        pub const ON_PRESSED: &str = "HotKey.onPressed";
    }
}

pub(crate) mod event {
    pub mod window {
        // Called when window has been properly initialized and can receive messages
        pub const INITIALIZE: &str = "event:Window.initialize";

        // Called when window became visible or hidden (boolean argument)
        pub const VISIBILITY_CHANGED: &str = "event:Window.visibilityChanged";

        // Delivered when user requested closing the window; Target window is responsible
        // for actually closing the window
        pub const CLOSE_REQUEST: &str = "event:Window.closeRequest";

        // Delivered when window is actually closed
        pub const CLOSE: &str = "event:Window.close";
    }
}

pub(crate) mod drag_data {
    pub mod key {
        pub const FILES: &str = "drag-data:internal:files";
        pub const URLS: &str = "drag-data:internal:urls";
    }
}
